### DISCUSSION TITLE
Possible cause for `object '.x' not found` error with `tar_combine`

### DISCUSSION THREAD

**Discussion by nicolaspayette** [ropensci/tarchetypes#126] [General]
I had a problem, for which I eventually found the cause, but it took me quite some time to figure it out, so I'm posting it here in case some else runs into the same issue.

Take the following example:

``` r
library(targets)
tar_script({
  library(tibble)
  library(tarchetypes)
  targets <- tar_map(
    values = lst(letter = c("A", "B", "C")),
    tar_target(x, tibble(value = letter))
  )
  lst(
    targets,
    tar_combine(y, targets, command = dplyr::bind_rows(!!!.x, .id = "name"))
  )
})

tar_manifest(fields = command)
#> Error:
#> ! Error running targets::tar_manifest()
#>   Target errors: targets::tar_meta(fields = error, complete_only = TRUE)
#>   Tips: https://books.ropensci.org/targets/debugging.html
#>   Last error: object '.x' not found

#> Backtrace:
#>     ▆
#>  1. └─targets::tar_manifest(fields = command)
#>  2.   └─targets:::callr_outer(...)
#>  3.     ├─targets:::if_any(...)
#>  4.     └─targets:::callr_error(condition = out, fun = fun)
#>  5.       └─targets::tar_throw_run(message)
#>  6.         └─targets::tar_error(...)
#>  7.           └─rlang::abort(message = message, class = class, call = tar_empty_envir)
```

<sup>Created on 2022-11-30 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

It fails with `object '.x' not found`. Can you see why?

The culprit is the `tibble::lst` function, which I have used here to build the target list. It doesn't work with the slicing operator `!!!.x`, presumably because it's doing something fancy under the hood.

Using plain old base `list` instead fixes the issue:

``` r
library(targets)
tar_script({
  library(tibble)
  library(tarchetypes)
  targets <- tar_map(
    values = lst(letter = c("A", "B", "C")),
    tar_target(x, tibble(value = letter))
  )
  list(
    targets,
    tar_combine(y, targets, command = dplyr::bind_rows(!!!.x, .id = "name"))
  )
})

tar_manifest(fields = command)
#> # A tibble: 4 × 2
#>   name  command                                                                 
#>   <chr> <chr>                                                                   
#> 1 x_B   "tibble(value = \"B\")"                                                 
#> 2 x_C   "tibble(value = \"C\")"                                                 
#> 3 x_A   "tibble(value = \"A\")"                                                 
#> 4 y     "dplyr::bind_rows(x_A = x_A, x_B = x_B, x_C = x_C, \n     .id = \"name\…
```

<sup>Created on 2022-11-30 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

I'm happy to carry on from here, but if someone could explain why `lst` doesn't play well with `!!!.x`, I'd be very curious.

**Comment by wlandau** [comment]
The problem is that `lst()` tries to tidy-evaluate `.x`. We instead want `lst()` to ignore `.x` and leave it for `tar_combine()`. This works:

```{r}
targets <- tar_map(
  values = lst(letter = c("A", "B", "C")),
  tar_target(x, tibble(value = letter))
)
combined <- tar_combine(y, targets, command = dplyr::bind_rows(!!!.x, .id = "name"))
lst(targets, combined)
```

**Reply by motorlearner** [reply]
Is there no way to solve this while having the `tar_combine` call inside `lst`?

When not specifying the `command` argument inside `tar_combine`, there is no `object .x not found` error (but a different one that asks to supply a `.name_spec` specification to `vctrs::vec_c`). 

When specifying the `command` argument, e.g. just specifying the default, the `object .x not found` error appears.

**Reply by wlandau** [reply]
> Is there no way to solve this while having the tar_combine call inside lst?

Is there a need to call `tibble::lst()` instead of `base::list()`?

**Reply by motorlearner** [reply]
No -- just I had already written it that way and don't want to re-run targets that take very long. 

For the record, it works fine when combining lists of tibbles (without specifying a `command`). So I just converted the relevant targets (nested lists to lists of tibbles).
